Makefile

```diff
@@ -1,29 +1,61 @@
-CC      = gcc
-CFLAGS  = -c -g -O2 -Wall
-LDFLAGS =
-
-SRC     = $(sort $(wildcard *.c))
-OBJ     = $(SRC:.c=.o)
+CC           = gcc
+CFLAGS       = -g -O2 -Wall
+LDFLAGS      = -L. -lmediacheck
+SHARED_FLAGS = -fPIC -fvisibility=hidden
 
+ARCH    := $(shell uname -m)
 GIT2LOG := $(shell if [ -x ./git2log ] ; then echo ./git2log --update ; else echo true ; fi)
 GITDEPS := $(shell [ -d .git ] && echo .git/HEAD .git/refs/heads .git/refs/tags)
 VERSION := $(shell $(GIT2LOG) --version VERSION ; cat VERSION)
 BRANCH  := $(shell [ -d .git ] && git branch | perl -ne 'print $$_ if s/^\*\s*//')
 PREFIX  := checkmedia-$(VERSION)
 
-%.o: %.c
-	$(CC) $(CFLAGS) -DVERSION=\"$(VERSION)\" -o $@ $<
+MAJOR_VERSION := $(shell $(GIT2LOG) --version VERSION ; cut -d . -f 1 VERSION)
+
+LIB_NAME     = libmediacheck
+LIB_FILENAME = $(LIB_NAME).so.$(VERSION)
+LIB_SONAME   = $(LIB_NAME).so.$(MAJOR_VERSION)
+
+DIGEST_SRC  = $(wildcard md5.c sha*.c)
+DIGEST_OBJ  = $(DIGEST_SRC:.c=.o)
+
+ifneq ($(filter x86_64, $(ARCH)),)
+LIBDIR  = /usr/lib64
+else
+LIBDIR  = /usr/lib
+endif
+
+all: checkmedia digestdemo
 
-all: checkmedia
+checkmedia: checkmedia.c $(LIB_FILENAME) mediacheck.h
+	$(CC) $(CFLAGS) $(LDFLAGS) checkmedia.c -DVERSION=\"$(VERSION)\" -o $@
 
-checkmedia: $(OBJ)
-	$(CC) $(OBJ) $(LDFLAGS) -o $@
+digestdemo: digestdemo.c $(LIB_FILENAME) mediacheck.h
+	$(CC) $(CFLAGS) $(LDFLAGS) digestdemo.c -o $@
+
+mediacheck.o: mediacheck.c mediacheck.h
+	$(CC) -c $(CFLAGS) $(SHARED_FLAGS) -o $@ $<
+
+$(DIGEST_OBJ): %.o: %.c %.h
+	$(CC) -c $(CFLAGS) $(SHARED_FLAGS) -o $@ $<
+
+$(LIB_FILENAME): $(DIGEST_OBJ) mediacheck.o
+	$(CC) -shared -Wl,-soname,$(LIB_SONAME) mediacheck.o $(DIGEST_OBJ) -o $(LIB_FILENAME)
+	@ln -snf $(LIB_FILENAME) $(LIB_SONAME)
+	@ln -snf $(LIB_SONAME) $(LIB_NAME).so
 
 changelog: $(GITDEPS)
 	$(GIT2LOG) --changelog changelog
 
+test: checkmedia
+	./testmediacheck
```

Comment:
naming is a bit inconsistent as it looks more like test of checkmedia and not mediacheck or am I wrong?

Reply:
it tests the mediacheck library

Reply:
ok